File: app/generate_icons.py
#!/usr/bin/env python3
"""각 팀별 아이콘 이미지 생성 (PNG → ICNS 변환)"""

import subprocess
import os
import tempfile
import struct
import zlib

DESKTOP = "/Users/sanghyunkim/Desktop"

TEAMS = [
    ("research",         "조사팀",       "🔍", (41, 128, 185)),
    ("debate",           "토론팀",       "💬", (142, 68, 173)),
    ("synthesis",        "종합팀",       "🧩", (39, 174, 96)),
    ("quality",          "퀄리티팀",     "✅", (22, 160, 133)),
    ("planning",         "기획팀",       "📋", (243, 156, 18)),
    ("design",           "디자인팀",     "🎨", (231, 76, 60)),
    ("development",      "개발팀",       "💻", (52, 73, 94)),
    ("devops",           "DevOps팀",     "⚙️", (127, 140, 141)),
    ("security",         "보안팀",       "🔒", (192, 57, 43)),
    ("marketing",        "마케팅팀",     "📢", (230, 126, 34)),
    ("sales",            "세일즈팀",     "🤝", (46, 204, 113)),
    ("customer-support", "고객지원팀",   "🎧", (52, 152, 219)),
    ("data-analytics",   "데이터분석팀", "📊", (155, 89, 182)),
    ("legal",            "법무팀",       "⚖️", (149, 165, 166)),
    ("finance",          "재무팀",       "💰", (241, 196, 15)),
    ("leadership",       "리더십팀",     "👔", (44, 62, 80)),
    ("documentation",    "문서화팀",     "📝", (26, 188, 156)),
    ("strategy",         "전략팀",       "🎯", (255, 107, 53)),
]


def create_png(width, height, emoji, label, bg_color):
    """Create a simple PNG icon using pure Python (no PIL needed)"""

    def make_png(w, h, pixels):
        """Generate PNG binary data from RGBA pixel array"""
        def chunk(chunk_type, data):
            c = chunk_type + data
            crc = struct.pack('>I', zlib.crc32(c) & 0xffffffff)
            return struct.pack('>I', len(data)) + c + crc

        header = b'\x89PNG\r\n\x1a\n'
        ihdr = chunk(b'IHDR', struct.pack('>IIBBBBB', w, h, 8, 6, 0, 0, 0))

        raw = b''
        for y in range(h):
            raw += b'\x00'  # filter byte
            for x in range(w):
                raw += bytes(pixels[y][x])

        idat = chunk(b'IDAT', zlib.compress(raw))
        iend = chunk(b'IEND', b'')
        return header + ihdr + idat + iend

    pixels = []
    r, g, b = bg_color
    corner_radius = width // 5

    for y in range(height):
        row = []
        for x in range(width):
            # Rounded rectangle
            in_rect = True
            alpha = 255

            # Check corners
            corners = [
                (corner_radius, corner_radius),
                (width - corner_radius - 1, corner_radius),
                (corner_radius, height - corner_radius - 1),
                (width - corner_radius - 1, height - corner_radius - 1),
            ]
            for cx, cy in corners:
                if ((x < corner_radius or x > width - corner_radius - 1) and
                    (y < corner_radius or y > height - corner_radius - 1)):
                    dx = x - cx
                    dy = y - cy
                    dist = (dx*dx + dy*dy) ** 0.5
                    if dist > corner_radius:
                        in_rect = False
                    elif dist > corner_radius - 1.5:
                        alpha = int(255 * (corner_radius - dist) / 1.5)

            if not in_rect:
                row.append((0, 0, 0, 0))
            else:
                # Gradient: lighter at top
                factor = 1.0 + 0.15 * (1.0 - y / height)
                pr = min(255, int(r * factor))
                pg = min(255, int(g * factor))
                pb = min(255, int(b * factor))
                row.append((pr, pg, pb, alpha))
        pixels.append(row)

    return make_png(width, height, pixels)


def create_iconset(emoji, label, bg_color, output_icns):
    """Create .icns file using iconutil"""
    with tempfile.TemporaryDirectory() as tmpdir:
        iconset_dir = os.path.join(tmpdir, "icon.iconset")
        os.makedirs(iconset_dir)

        sizes = {
            "icon_16x16.png": 16,
            "icon_16x16@2x.png": 32,
            "icon_32x32.png": 32,
            "icon_32x32@2x.png": 64,
            "icon_128x128.png": 128,
            "icon_128x128@2x.png": 256,
            "icon_256x256.png": 256,
            "icon_256x256@2x.png": 512,
            "icon_512x512.png": 512,
            "icon_512x512@2x.png": 1024,
        }

        for filename, size in sizes.items():
            png_data = create_png(size, size, emoji, label, bg_color)
            png_path = os.path.join(iconset_dir, filename)
            with open(png_path, 'wb') as f:
                f.write(png_data)

        # Use sips + iconutil to create proper icon with text overlay
        # First create a master 1024px PNG with text using Core Graphics via osascript
        master_png = os.path.join(tmpdir, "master.png")
        create_icon_with_text(1024, emoji, label, bg_color, master_png)

        if os.path.exists(master_png):
            # Replace all sizes with properly rendered versions
            for filename, size in sizes.items():
                sized_png = os.path.join(iconset_dir, filename)
                subprocess.run([
                    "sips", "-z", str(size), str(size),
                    master_png, "--out", sized_png
                ], capture_output=True)

        subprocess.run(["iconutil", "-c", "icns", iconset_dir, "-o", output_icns],
                       capture_output=True)


def create_icon_with_text(size, emoji, label, bg_color, output_path):
    """Use macOS native tools to create icon with emoji and text"""
    r, g, b = bg_color
    rf, gf, bf = r/255, g/255, b/255

    # Use Objective-C via Python to render
    script = f'''
    tell application "System Events"
    end tell

    use framework "AppKit"
    use framework "Foundation"
    use scripting additions

    set imgSize to {size}
    set theImage to current application's NSImage's alloc()'s initWithSize:{{imgSize, imgSize}}
    theImage's lockFocus()

    -- Background rounded rect
    set bgColor to current application's NSColor's colorWithRed:{rf} green:{gf} blue:{bf} alpha:1.0
    set bgPath to current application's NSBezierPath's bezierPathWithRoundedRect:{{{{0, 0}}, {{imgSize, imgSize}}}} xRadius:(imgSize * 0.2) yRadius:(imgSize * 0.2)
    bgColor's setFill()
    bgPath's fill()

    -- Lighter overlay at top
    set gradColor to current application's NSColor's colorWithWhite:1.0 alpha:0.15
    set gradRect to current application's NSBezierPath's bezierPathWithRect:{{{{0, imgSize * 0.5}}, {{imgSize, imgSize * 0.5}}}}
    gradColor's setFill()
    gradRect's fill()

    -- Emoji text
    set emojiSize to imgSize * 0.45
    set emojiFont to current application's NSFont's systemFontOfSize:emojiSize
    set emojiAttrs to current application's NSDictionary's dictionaryWithObjects:{{emojiFont}} forKeys:{{current application's NSFontAttributeName}}
    set emojiStr to current application's NSAttributedString's alloc()'s initWithString:"{emoji}" attributes:emojiAttrs
    set emojiStrSize to emojiStr's |size|()
    set emojiX to (imgSize - (emojiStrSize's width)) / 2
    set emojiY to imgSize * 0.35
    emojiStr's drawAtPoint:{{emojiX, emojiY}}

    -- Label text
    set labelSize to imgSize * 0.13
    set labelFont to current application's NSFont's boldSystemFontOfSize:labelSize
    set labelColor to current application's NSColor's whiteColor()
    set shadowObj to current application's NSShadow's alloc()'s init()
    shadowObj's setShadowOffset:{{0, -1}}
    shadowObj's setShadowBlurRadius:3
    shadowObj's setShadowColor:(current application's NSColor's colorWithWhite:0.0 alpha:0.5)
    set labelAttrs to current application's NSDictionary's dictionaryWithObjects:{{labelFont, labelColor, shadowObj}} forKeys:{{current application's NSFontAttributeName, current application's NSForegroundColorAttributeName, current application's NSShadowAttributeName}}
    set labelStr to current application's NSAttributedString's alloc()'s initWithString:"{label}" attributes:labelAttrs
    set labelStrSize to labelStr's |size|()
    set labelX to (imgSize - (labelStrSize's width)) / 2
    set labelY to imgSize * 0.12
    labelStr's drawAtPoint:{{labelX, labelY}}

    theImage's unlockFocus()

    -- Save as PNG
    set tiffData to theImage's TIFFRepresentation()
    set bitmapRep to current application's NSBitmapImageRep's imageRepWithData:tiffData
    set pngData to bitmapRep's representationUsingType:(current application's NSBitmapImageFileTypePNG) properties:(current application's NSDictionary's dictionary())
    pngData's writeToFile:"{output_path}" atomically:true
    '''

    subprocess.run(["osascript", "-l", "AppleScript", "-e", script],
                   capture_output=True, timeout=30)


def main():
    for key, kr_name, emoji, color in TEAMS:
        clean_emoji = emoji.replace('\ufe0f', '')  # remove variation selector
        app_name = f"{clean_emoji} {kr_name}"
        app_path = os.path.join(DESKTOP, f"{app_name}.app")
        icns_path = os.path.join(app_path, "Contents", "Resources", "AppIcon.icns")

        if not os.path.isdir(app_path):
            print(f"⚠️  {app_name}.app not found, skipping")
            continue

        print(f"🎨 Generating icon for {kr_name}...", end=" ", flush=True)
        create_iconset(clean_emoji, kr_name, color, icns_path)

        if os.path.exists(icns_path):
            print("✅")
        else:
            # Fallback: use the pure Python PNG-based icon
            print("(fallback) ", end="", flush=True)
            with tempfile.TemporaryDirectory() as tmpdir:
                iconset_dir = os.path.join(tmpdir, "icon.iconset")
                os.makedirs(iconset_dir)
                sizes = {
                    "icon_16x16.png": 16, "icon_16x16@2x.png": 32,
                    "icon_32x32.png": 32, "icon_32x32@2x.png": 64,
                    "icon_128x128.png": 128, "icon_128x128@2x.png": 256,
                    "icon_256x256.png": 256, "icon_256x256@2x.png": 512,
                    "icon_512x512.png": 512, "icon_512x512@2x.png": 1024,
                }
                for fname, sz in sizes.items():
                    png_data = create_png(sz, sz, clean_emoji, kr_name, color)
                    with open(os.path.join(iconset_dir, fname), 'wb') as f:
                        f.write(png_data)
                subprocess.run(["iconutil", "-c", "icns", iconset_dir, "-o", icns_path],
                               capture_output=True)
            print("✅" if os.path.exists(icns_path) else "❌")

    # Clear icon cache
    print("\n🔄 Refreshing icon cache...")
    subprocess.run(["touch", *[
        os.path.join(DESKTOP, f"{e.replace(chr(0xfe0f),'')} {n}.app")
        for _, n, e, _ in TEAMS
    ]], capture_output=True)

    print("\n✨ 모든 팀 아이콘 생성 완료!")


if __name__ == "__main__":
    main()

File: app/generate_icons.swift
import AppKit
import Foundation

struct TeamIcon {
    let key: String
    let name: String
    let emoji: String
    let subIcons: [String]
    let primary: (CGFloat, CGFloat, CGFloat)
    let secondary: (CGFloat, CGFloat, CGFloat)
    let accent: (CGFloat, CGFloat, CGFloat)
}

let teams: [TeamIcon] = [
    TeamIcon(key: "research",         name: "조사팀",       emoji: "🔍", subIcons: ["📡","🌐","📑"], primary: (0.16, 0.50, 0.73), secondary: (0.20, 0.60, 0.86), accent: (0.68, 0.84, 0.95)),
    TeamIcon(key: "debate",           name: "토론팀",       emoji: "💬", subIcons: ["⚡","🗣️","🎯"], primary: (0.56, 0.27, 0.68), secondary: (0.61, 0.35, 0.71), accent: (0.84, 0.74, 0.89)),
    TeamIcon(key: "synthesis",        name: "종합팀",       emoji: "🧩", subIcons: ["📐","🔗","📊"], primary: (0.15, 0.68, 0.38), secondary: (0.18, 0.80, 0.44), accent: (0.67, 0.92, 0.78)),
    TeamIcon(key: "quality",          name: "퀄리티팀",     emoji: "✅", subIcons: ["🔬","🛡️","⭐"], primary: (0.09, 0.63, 0.52), secondary: (0.10, 0.74, 0.61), accent: (0.64, 0.89, 0.84)),
    TeamIcon(key: "planning",         name: "기획팀",       emoji: "📋", subIcons: ["🗓️","🎯","📌"], primary: (0.95, 0.61, 0.07), secondary: (0.95, 0.77, 0.06), accent: (0.98, 0.91, 0.62)),
    TeamIcon(key: "design",           name: "디자인팀",     emoji: "🎨", subIcons: ["✏️","🖌️","💎"], primary: (0.91, 0.30, 0.24), secondary: (0.93, 0.44, 0.39), accent: (0.96, 0.72, 0.69)),
    TeamIcon(key: "development",      name: "개발팀",       emoji: "💻", subIcons: ["🔧","⚡","🚀"], primary: (0.17, 0.24, 0.31), secondary: (0.20, 0.29, 0.37), accent: (0.52, 0.57, 0.62)),
    TeamIcon(key: "devops",           name: "DevOps팀",     emoji: "⚙️", subIcons: ["🐳","☁️","🔄"], primary: (0.38, 0.42, 0.42), secondary: (0.50, 0.55, 0.55), accent: (0.70, 0.73, 0.73)),
    TeamIcon(key: "security",         name: "보안팀",       emoji: "🔒", subIcons: ["🛡️","🔑","🚨"], primary: (0.66, 0.20, 0.15), secondary: (0.75, 0.22, 0.17), accent: (0.90, 0.69, 0.67)),
    TeamIcon(key: "marketing",        name: "마케팅팀",     emoji: "📢", subIcons: ["📈","🎪","🚀"], primary: (0.90, 0.49, 0.13), secondary: (0.95, 0.61, 0.07), accent: (0.97, 0.77, 0.44)),
    TeamIcon(key: "sales",            name: "세일즈팀",     emoji: "🤝", subIcons: ["💼","📞","💎"], primary: (0.15, 0.68, 0.38), secondary: (0.35, 0.84, 0.55), accent: (0.67, 0.92, 0.78)),
    TeamIcon(key: "customer-support", name: "고객지원팀",   emoji: "🎧", subIcons: ["💝","📱","🌟"], primary: (0.20, 0.60, 0.86), secondary: (0.36, 0.68, 0.89), accent: (0.68, 0.84, 0.95)),
    TeamIcon(key: "data-analytics",   name: "데이터분석팀", emoji: "📊", subIcons: ["🧮","📈","🔮"], primary: (0.56, 0.27, 0.68), secondary: (0.69, 0.48, 0.77), accent: (0.84, 0.74, 0.89)),
    TeamIcon(key: "legal",            name: "법무팀",       emoji: "⚖️", subIcons: ["📜","🏛️","🔍"], primary: (0.39, 0.43, 0.45), secondary: (0.50, 0.55, 0.55), accent: (0.70, 0.73, 0.73)),
    TeamIcon(key: "finance",          name: "재무팀",       emoji: "💰", subIcons: ["📉","💳","🏦"], primary: (0.83, 0.67, 0.05), secondary: (0.95, 0.77, 0.06), accent: (0.98, 0.91, 0.62)),
    TeamIcon(key: "leadership",       name: "리더십팀",     emoji: "👔", subIcons: ["⭐","🏆","🎖️"], primary: (0.09, 0.13, 0.16), secondary: (0.17, 0.24, 0.31), accent: (0.36, 0.43, 0.49)),
    TeamIcon(key: "documentation",    name: "문서화팀",     emoji: "📝", subIcons: ["📖","🗂️","✍️"], primary: (0.08, 0.56, 0.47), secondary: (0.10, 0.74, 0.61), accent: (0.64, 0.89, 0.84)),
    TeamIcon(key: "strategy",         name: "전략팀",       emoji: "🎯", subIcons: ["🧭","📐","🚀"], primary: (1.00, 0.42, 0.21), secondary: (1.00, 0.55, 0.35), accent: (1.00, 0.78, 0.67)),
]

let desktop = "/Users/sanghyunkim/Desktop"
let imgSize: CGFloat = 1024

func generateIcon(team: TeamIcon, outputPath: String) {
    let image = NSImage(size: NSSize(width: imgSize, height: imgSize))
    image.lockFocus()

    let cornerR = imgSize * 0.22

    // 1. Base background
    let bgColor = NSColor(red: team.primary.0, green: team.primary.1, blue: team.primary.2, alpha: 1.0)
    let bgPath = NSBezierPath(roundedRect: NSRect(x: 0, y: 0, width: imgSize, height: imgSize),
                              xRadius: cornerR, yRadius: cornerR)
    bgColor.setFill()
    bgPath.fill()

    // 2. Secondary gradient overlay (top half)
    let secColor = NSColor(red: team.secondary.0, green: team.secondary.1, blue: team.secondary.2, alpha: 0.5)
    let secPath = NSBezierPath(roundedRect: NSRect(x: 0, y: imgSize * 0.4, width: imgSize, height: imgSize * 0.6),
                               xRadius: cornerR, yRadius: cornerR)
    secColor.setFill()
    secPath.fill()

    // 3. Decorative accent circle (top-right)
    let accentColor = NSColor(red: team.accent.0, green: team.accent.1, blue: team.accent.2, alpha: 0.12)
    let circleSize = imgSize * 0.7
    let circlePath = NSBezierPath(ovalIn: NSRect(x: imgSize * 0.45, y: imgSize * 0.35, width: circleSize, height: circleSize))
    accentColor.setFill()
    circlePath.fill()

    // 4. Bottom dark strip
    let stripColor = NSColor(white: 0.0, alpha: 0.2)
    let stripPath = NSBezierPath(roundedRect: NSRect(x: 0, y: 0, width: imgSize, height: imgSize * 0.28),
                                 xRadius: cornerR, yRadius: 0)
    stripColor.setFill()
    stripPath.fill()

    // 5. Sub icons (decorative, semi-transparent)
    let subFont = NSFont.systemFont(ofSize: imgSize * 0.11)
    let subColor = NSColor.white.withAlphaComponent(0.3)
    let subAttrs: [NSAttributedString.Key: Any] = [.font: subFont, .foregroundColor: subColor]
    let positions: [(CGFloat, CGFloat)] = [(0.08, 0.82), (0.83, 0.78), (0.80, 0.42)]
    for (i, sub) in team.subIcons.prefix(3).enumerated() {
        let str = NSAttributedString(string: sub, attributes: subAttrs)
        str.draw(at: NSPoint(x: imgSize * positions[i].0, y: imgSize * positions[i].1))
    }

    // 6. Main emoji (large, centered)
    let emojiFont = NSFont.systemFont(ofSize: imgSize * 0.42)
    let emojiAttrs: [NSAttributedString.Key: Any] = [.font: emojiFont]
    let emojiStr = NSAttributedString(string: team.emoji, attributes: emojiAttrs)
    let emojiStrSize = emojiStr.size()
    emojiStr.draw(at: NSPoint(x: (imgSize - emojiStrSize.width) / 2, y: imgSize * 0.33))

    // 7. Team name with shadow
    let labelFont = NSFont.boldSystemFont(ofSize: imgSize * 0.115)
    let shadow = NSShadow()
    shadow.shadowOffset = NSSize(width: 0, height: -2)
    shadow.shadowBlurRadius = 8
    shadow.shadowColor = NSColor.black.withAlphaComponent(0.7)
    let labelAttrs: [NSAttributedString.Key: Any] = [
        .font: labelFont,
        .foregroundColor: NSColor.white,
        .shadow: shadow
    ]
    let labelStr = NSAttributedString(string: team.name, attributes: labelAttrs)
    let labelSize = labelStr.size()
    labelStr.draw(at: NSPoint(x: (imgSize - labelSize.width) / 2, y: imgSize * 0.08))

    // 8. Thin white border
    let borderColor = NSColor.white.withAlphaComponent(0.12)
    let borderPath = NSBezierPath(roundedRect: NSRect(x: 1, y: 1, width: imgSize - 2, height: imgSize - 2),
                                  xRadius: cornerR, yRadius: cornerR)
    borderPath.lineWidth = 3
    borderColor.setStroke()
    borderPath.stroke()

    image.unlockFocus()

    // Save
    guard let tiffData = image.tiffRepresentation,
          let bitmapRep = NSBitmapImageRep(data: tiffData),
          let pngData = bitmapRep.representation(using: .png, properties: [:]) else {
        print("  ❌ \(team.name) - render failed")
        return
    }
    try! pngData.write(to: URL(fileURLWithPath: outputPath))
}

func pngToIcns(masterPng: String, icnsPath: String) -> Bool {
    let tmpDir = NSTemporaryDirectory() + "iconset_\(UUID().uuidString)"
    let iconsetDir = tmpDir + "/icon.iconset"

    try! FileManager.default.createDirectory(atPath: iconsetDir, withIntermediateDirectories: true)

    let sizes: [(String, Int)] = [
        ("icon_16x16.png", 16), ("icon_16x16@2x.png", 32),
        ("icon_32x32.png", 32), ("icon_32x32@2x.png", 64),
        ("icon_128x128.png", 128), ("icon_128x128@2x.png", 256),
        ("icon_256x256.png", 256), ("icon_256x256@2x.png", 512),
        ("icon_512x512.png", 512), ("icon_512x512@2x.png", 1024),
    ]

    for (fname, sz) in sizes {
        let outPath = iconsetDir + "/" + fname
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/usr/bin/sips")
        p.arguments = ["-z", "\(sz)", "\(sz)", masterPng, "--out", outPath]
        p.standardOutput = FileHandle.nullDevice
        p.standardError = FileHandle.nullDevice
        try! p.run()
        p.waitUntilExit()
    }

    let p2 = Process()
    p2.executableURL = URL(fileURLWithPath: "/usr/bin/iconutil")
    p2.arguments = ["-c", "icns", iconsetDir, "-o", icnsPath]
    p2.standardOutput = FileHandle.nullDevice
    p2.standardError = FileHandle.nullDevice
    try! p2.run()
    p2.waitUntilExit()

    try? FileManager.default.removeItem(atPath: tmpDir)
    return FileManager.default.fileExists(atPath: icnsPath)
}

// Main
print("=======================================================")
print("  팀별 고유 앱 아이콘 생성 (Swift)")
print("=======================================================")
print()

var success = 0
var fail = 0

for team in teams {
    // Find .app
    let contents = try! FileManager.default.contentsOfDirectory(atPath: desktop)
    guard let appEntry = contents.first(where: { $0.hasSuffix(".app") && $0.contains(team.name) }) else {
        print("  ❌ \(team.name) - .app not found")
        fail += 1
        continue
    }
    let appPath = desktop + "/" + appEntry
    let icnsPath = appPath + "/Contents/Resources/AppIcon.icns"

    print("  🎨 \(team.emoji) \(team.name)...", terminator: " ")
    fflush(stdout)

    let tmpPng = NSTemporaryDirectory() + "team_icon_\(team.key).png"
    generateIcon(team: team, outputPath: tmpPng)

    if FileManager.default.fileExists(atPath: tmpPng) && pngToIcns(masterPng: tmpPng, icnsPath: icnsPath) {
        let attrs = try! FileManager.default.attributesOfItem(atPath: icnsPath)
        let size = attrs[.size] as! Int
        print("✅ (\(size.formatted()) bytes)")
        success += 1
    } else {
        print("❌ failed")
        fail += 1
    }

    try? FileManager.default.removeItem(atPath: tmpPng)
}

// Refresh
print()
print("  🔄 아이콘 캐시 갱신...")
for team in teams {
    let contents = try! FileManager.default.contentsOfDirectory(atPath: desktop)
    if let appEntry = contents.first(where: { $0.hasSuffix(".app") && $0.contains(team.name) }) {
        let p = Process()
        p.executableURL = URL(fileURLWithPath: "/usr/bin/touch")
        p.arguments = [desktop + "/" + appEntry]
        try! p.run()
        p.waitUntilExit()
    }
}

print()
print("=======================================================")
print("  결과: \(success)개 성공 / \(fail)개 실패")
print("=======================================================")
